spec/requests/classified_listings_spec.rb

```diff
@@ -272,5 +326,140 @@
         expect(org_listing.reload.bumped_at >= previous_bumped_at).to eq(true)
       end
     end
+
+    context "when the publish action is called" do
+      let(:params) { { classified_listing: { action: "publish" } } }
+
+      it "publishes a draft and charges user credits if first publish" do
+        cost = ClassifiedListing.cost_by_category(listing_draft.category)
+        create_list(:credit, cost, user: user)
+        expect do
+          put "/listings/#{listing_draft.id}", params: params
+        end.to change(user.credits.spent, :size).by(cost)
+        expect(listing_draft.reload.published).to eq(true)
+      end
+
+      it "publishes a draft and charges org credits before user credits if first publish" do
+        cost = ClassifiedListing.cost_by_category(org_listing_draft.category)
+        create_list(:credit, cost, organization: organization)
+        expect do
+          put "/listings/#{org_listing_draft.id}", params: params
+        end.to change(organization.credits.spent, :size).by(cost)
+        expect(org_listing_draft.reload.published).to eq(true)
+      end
+
+      it "publishes a draft that has already been charged and is within 30 days of bump date" do
+        listing.published = false
+        listing.save
+        expect do
+          put "/listings/#{listing.id}", params: params
+        end.to change(user.credits.spent, :size).by(0)
+        expect(listing.reload.published).to eq(true)
+      end
+
+      it "fails to publish draft with insufficient credits" do
+        expect do
+          put "/listings/#{listing_draft.id}", params: params
+        end.to change(user.credits.spent, :size).by(0)
+        expect(listing_draft.reload.published).to eq(false)
+      end
+    end
+
+    context "when the unpublish action is called" do
+      let(:params) { { classified_listing: { action: "unpublish" } } }
+
+      it "unpublishes a published listing" do
+        put "/listings/#{listing.id}", params: params
+        expect(listing.reload.published).to eq(false)
+      end
+    end
+  end
+
+  describe "DEL /listings/:id" do
+    let(:listing) { create(:classified_listing, user: user) }
+    let(:listing_draft) { create(:classified_listing, user: user) }
+    let(:organization) { create(:organization) }
+    let(:org_listing) { create(:classified_listing, user: user, organization: organization) }
+    let(:org_listing_draft) { create(:classified_listing, user: user, organization: organization) }
+
+    before do
+      sign_in user
+      listing.save
+      org_listing.save
+      listing_draft.bumped_at = nil
```

Comment:
Why not provide `bumped_at` and `published` values on create? :thinking: 
You can also use `let!` instead of `let` if you need to use the record in all (many) of the examples. Some devs are against it, but it's cleaner than `let` + `before` + `save` anyway (and I don't see anything bad in using `let!` in such cases). Same applies to listings `save` in the previous lines.